### PR TITLE
Transform layered styles

### PR DESCRIPTION
**Pull request by romainmenke**
Done :

- remove `@layer` and adjust specificity of selectors
- remove `@layer` and re-order `@keyframes`, `@counter-style`, ...
- some refactoring

Known issues :

- unable to handle layers having a different order in multiple `@media` rules
- `!important` behaviour is lacking WPT tests and browser implementations are inconsistent

Not handled here :
- `revert`
- `revert-layer`

### CODE REVIEW

File: plugins/postcss-cascade-layers/src/compound-selector-order.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import selectorParser from 'postcss-selector-parser';
```
Comment (romainmenke) [inline plugins/postcss-cascade-layers/src/compound-selector-order.ts:1]: This file and functions therein allow us to concatenate any selector with `:not()` into a non-broken selector.

This prevents issues that would arise from `.foo::before` + `:not()`

File: plugins/postcss-cascade-layers/src/desugar-nested-layers.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import type { Container, AtRule, ChildNode } from 'postcss';
```
Comment (romainmenke) [inline plugins/postcss-cascade-layers/src/desugar-nested-layers.ts:1]:
```css
@layer A {
  @layer B {
    .foo { color: green; }
  }
}
```

Is equivalent to :

```css
@layer A.B {
  .foo { color: green; }
}
```

Desugaring to `@layer A.B` simplifies further processing.

**Comment by romainmenke** [comment]
@oluoluoxenfree This PR contains a big chunk of work towards completing this plugin :)
Please ask any questions you might have about this change.

File: plugins/postcss-cascade-layers/src/specificity.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
+import selectorParser from 'postcss-selector-parser';
```
Comment (sanajaved7) [inline plugins/postcss-cascade-layers/src/specificity.ts]: @romainmenke Thanks for all your work on this!! I noticed this `selectorSpecificity` method seems to be used in other plugins as well, would it make sense for this to be provided centrally from postcss? Also, this code was a bit difficult to follow. Would you be able to clarify what `a`, `b`, and `c` specifically mean in relation to the specificity? I wasn't sure for example how `a` is associated with `id` and `c` with `tag`. Thanks again!

File: plugins/postcss-cascade-layers/test/atrules.css [hunk 4]
```diff
@@ -0,0 +1,43 @@
+@layer framework, override;
```
Comment (sanajaved7) [inline plugins/postcss-cascade-layers/test/atrules.css:1]: Great tests! 🎉

File: plugins/postcss-cascade-layers/src/specificity.ts (hunk 3 above)
Comment (romainmenke) [reply plugins/postcss-cascade-layers/src/specificity.ts]: > I noticed this `selectorSpecificity` method seems to be used in other plugins as well, would it make sense for this to be provided centrally from postcss

Ideally we would have a way (a utilities package?) to share code across multiple postcss plugins. But as we still actively support node 12 and 14, we are careful with the mono-repo structure. At the moment we feel that copy/paste re-use adds the least complexity.

Once node 20 reaches lts and each version (16, 18, 20) has workspaces support this will change. It will then become much safer and easier to share code in a way to works well during development and in CI.

> Would you be able to clarify what a, b, and c specifically mean in relation to the specificity?

https://www.w3.org/TR/selectors-4/#specificity-rules

`a`, `b` and `c` are taken from the CSS specification.
We followed their naming convention when writing this code.

It is indeed hard to follow and this is problematic for contributors. Thank you for pointing this out!

Is it ok if we backlog this and tackle this in a specific issue? Some comments and references to the specification should be added, but this can best be done for all plugins at once.

Comment (sanajaved7) [reply plugins/postcss-cascade-layers/src/specificity.ts]: @romainmenke thanks for the helpful context and yes absolutely, this can be done at a later time, and thanks again! I appreciate your help!

Comment (romainmenke) [reply plugins/postcss-cascade-layers/src/specificity.ts]: @sanajaved7 We took the time today to ship a package dedicated to the specificity calculations : https://www.npmjs.com/package/@csstools/selector-specificity

The code here was actually mature enough and having a dedicated package makes it easier to have better tests specifically for this code.

> At the moment we feel that copy/paste re-use adds the least complexity.

With it being used by 4 plugins and these being the most complex, it is better to de-dupe it now and have the extra safety of the better test coverage :)

Thank you for bringing my attention to this part.
Always good to re-evaluate choices like these from time to time.

File: plugins/postcss-cascade-layers/src/desugar-and-parse-layer-names.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	root.walkAtRules('layer', (layerRule) => {
+		if (layerRule.params) {
+			const layerNameList: Array<string> = [];
+			let isInvalidLayerName = false;
+
+			selectorParser().astSync(layerRule.params).each((selector) => {
+				const currentLayerNameParts: Array<string> = [];
+
+				selector.walk((node) => {
+					switch (node.type) {
```
Comment (oluoluoxenfree) [inline plugins/postcss-cascade-layers/src/desugar-and-parse-layer-names.ts:23]: what do the node types mean?

File: plugins/postcss-cascade-layers/src/desugar-nested-layers.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+		root.walkAtRules('layer', (layerRule) => {
+			if (layerRule.parent === root) {
+				return;
+			}
+
+			if (layerRule.parent.type === 'atrule' && (layerRule.parent as AtRule).name === 'layer') {
+				const parent = layerRule.parent as AtRule;
+
+				model.layerNameParts.set(`${parent.params}.${layerRule.params}`, [...model.layerNameParts.get(parent.params), ...model.layerNameParts.get(layerRule.params)]);
```
Comment (oluoluoxenfree) [inline plugins/postcss-cascade-layers/src/desugar-nested-layers.ts:22]: what is this line doing?

**Review by oluoluoxenfree — COMMENTED** [review]
What are the next steps on this also? 

Thanks so much for all your work on this @romainmenke !

File: plugins/postcss-cascade-layers/src/desugar-and-parse-layer-names.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+export function desugarAndParseLayerNames(root: Container, model: Model) {
+	// - parse layer names
+	// - rename anon layers
+	// - handle empty layers
+	root.walkAtRules('layer', (layerRule) => {
+		if (layerRule.params) {
+			const layerNameList: Array<string> = [];
+			let isInvalidLayerName = false;
+
+			selectorParser().astSync(layerRule.params).each((selector) => {
```
Comment (romainmenke) [inline plugins/postcss-cascade-layers/src/desugar-and-parse-layer-names.ts:19]:
```suggestion
			// Layer names ("A.B, C") have a similar syntax as selector lists with only class and tag name selectors.
			// - comma separated
			// - "." as a delimiter between idents.
			// We can use "postcss-selector-parser" to analyse layer names so that we don't have to implement our own tokenizer and parser.
			selectorParser().astSync(layerRule.params).each((selector) => {
```

File: plugins/postcss-cascade-layers/src/desugar-nested-layers.ts (hunk 6 above)
Comment (romainmenke) [inline plugins/postcss-cascade-layers/src/desugar-nested-layers.ts:22]:
```suggestion
				// Concatenate the current layer params with those of the parent. Store the result in the data model.
								model.layerNameParts.set(`${parent.params}.${layerRule.params}`, [...model.layerNameParts.get(parent.params), ...model.layerNameParts.get(layerRule.params)]);
```

Comment (romainmenke) [reply plugins/postcss-cascade-layers/src/desugar-nested-layers.ts:21]: Here we concatenate the parent and current layer params for these cases :

```css
@layer A {
  @layer C.D {}
}

@layer A.B {
  @layer C {}
}

@layer A.B {
  @layer C.D {}
}
```

We do not want to re-parse these but we still want the combined information.

Comment (romainmenke) [reply plugins/postcss-cascade-layers/src/desugar-nested-layers.ts:22]: Here we concatenate the parent and current layer params for these cases :

```css
@layer A {
  @layer C.D {}
}

@layer A.B {
  @layer C {}
}

@layer A.B {
  @layer C.D {}
}
```

We do not want to re-parse these but we still want the combined information.

**Comment by romainmenke** [comment]
@oluoluoxenfree I've added some suggestions for code comments that hopefully answer your questions. Please let me know if these are not yet fully clear.

> What are the next steps on this also?

If this batch of changes is clear and works for you then it can be merged.

In general I think only `!important` and `revert|revert-layer` still need to be handled.

File: plugins/postcss-cascade-layers/src/specificity.ts (hunk 3 above)
Comment (sanajaved7) [reply plugins/postcss-cascade-layers/src/specificity.ts]: @romainmenke that is so awesome! Thank you for taking the time to do that and in general, for always being so open to feedback! 🙏🏼 💯

**Review by oluoluoxenfree — APPROVED** [review]

**Comment by oluoluoxenfree** [comment]
@romainmenke is this ready to merge?

**Comment by romainmenke** [comment]
Fully ready on my end :)
Go ahead!
